### PR TITLE
added basic batch norm functionality and fixed compute_cose and L_mod…

### PR DESCRIPTION
**Pull request by galbraun**
…el_forward

### CODE REVIEW

File: assignment1/fc_net.py [hunk 1]
```diff
@@ -54,12 +54,30 @@ def L_model_backward(self, AL, Y, caches):
...
         """
-        pass
+
+        grads = []
+
+        last_layer_dA = -(Y/AL) + ((1-Y)/1-AL)
+        grads['dA' + str(self.num_layers - 1)] = last_layer_dA
+
+        dA, dW, db = linear_activation_backward(last_layer_dA, caches[self.num_layers - 1], 'sigmoid')
+        grads['dW' + str(self.num_layers - 1)] = dW
```
Comment (elisim) [inline assignment1/fc_net.py]: move self.num_layers - 1 to variable

File: assignment1/fc_net.py [hunk 2]
```diff
@@ -54,12 +54,30 @@ def L_model_backward(self, AL, Y, caches):
...
+        :param AL: the probabilities vector, the output of the forward propagation (L_model_forward)
+        :param Y: the true labels vector (the "ground truth" - true classifications)
+        :param caches: list of caches containing for each layer: a) the linear cache; b) the activation cache
+        :return: a dictionary with the gradients
         """
-        pass
+
+        grads = []
+
+        last_layer_dA = -(Y/AL) + ((1-Y)/1-AL)
```
Comment (elisim) [inline assignment1/fc_net.py]: that is cross entropy grad? 
maybe add comment?

File: assignment1/fc_net.py [hunk 3]
```diff
@@ -54,12 +54,30 @@ def L_model_backward(self, AL, Y, caches):
...
-        :param caches:
-        :return:
+        :param AL: the probabilities vector, the output of the forward propagation (L_model_forward)
+        :param Y: the true labels vector (the "ground truth" - true classifications)
+        :param caches: list of caches containing for each layer: a) the linear cache; b) the activation cache
+        :return: a dictionary with the gradients
         """
-        pass
+
+        grads = []
```
[the remark this answers is not in the document]
Comment (galbraun) [reply assignment1/fc_net.py]: Fuck you are right ❤️

File: assignment1/fc_net.py (hunk 1 above)
Comment (galbraun) [reply assignment1/fc_net.py]: 👍

File: assignment1/fc_net.py (hunk 2 above)
Comment (galbraun) [reply assignment1/fc_net.py]: Look in the Lecture_1.pptx slide 50. i'll add a comment.

Comment (elisim) [reply assignment1/fc_net.py]: Okay, I don't understand it. That is the loss gradient?  what is the loss in lecture 1?
 By the way, It should be (1-Y)/(1-AL)

File: assignment1/fc_net.py [hunk 4]
```diff
@@ -80,8 +80,24 @@ def L_model_backward(self, AL, Y, caches):
...
+        Updates parameters using gradient descen
+
+        :param parameters: a python dictionary containing the DNN architecture’s parameters
+        :param grads: a python dictionary containing the gradients (generated by L_model_backward)
+        :param learning_rate: the learning rate used to update the parameters (the “alpha”)
+        :return: – the updated values of the parameters object provided as input
+        """
+
+        for layer_idx in range(self.num_layers):
+            old_W, dW = parameters['W' + str(layer_idx)], grads['db' + str(layer_idx)]
```
Comment (elisim) [inline assignment1/fc_net.py]: layers idx start from 1. should be layer_idx + 1

File: assignment1/layers.py [hunk 5]
```diff
@@ -93,8 +93,8 @@ def linear_backward(dZ, cache):
     A_prev_reshaped = A_prev.reshape(N, -1)
 
     dA_prev = dZ.dot(W.T).reshape(A_prev.shape)
-    dW = A_prev_reshaped.T.dot(dZ)
-    db = np.sum(dZ, axis=0)
+    dW = A_prev_reshaped.T.dot(dZ) / N
```
Comment (elisim) [inline assignment1/layers.py]: can you share a link where you read that it is necessary? 
I could not find it

File: assignment1/fc_net.py [hunk 6]
```diff
@@ -99,6 +99,46 @@ def update_parameters(self, parameters, grads, learning_rate):
...
+                                        function (calculated by the compute_cost function). One value is to be saved
+                                        after each 100 training iterations (e.g. 3000 iterations -> 30 values)..
+        """
+        # initialization
+        parameters = initialize_parameters(layers_dims)
+        costs = []
+
+        for i in range(num_iterations):
+            # choose the batch
+            batch_idx = np.random.choice(np.arange(X.shape[1]), size=batch_size, replace=False)
```
Comment (elisim) [inline assignment1/fc_net.py]: wrong, you can chose the same batch twice. 

it should be something like 

```
def next_batch(X, y, batchSize):
	# loop over our dataset `X` in mini-batches of size `batchSize`
	for i in np.arange(0, X.shape[1], batchSize):
		# yield a tuple of the current batched data and labels
		yield (X[i:i + batchSize], y[i:i + batchSize])
```

File: assignment1/fc_net.py [hunk 7]
```diff
@@ -99,6 +99,46 @@ def update_parameters(self, parameters, grads, learning_rate):
...
+        :param learning_rate: the learning rate
+        :param num_iterations: number of iterations
+        :param batch_size: the number of examples in a single training batch.
+        :return: (parameters, costs) - the parameters learnt by the system during the training (the same parameters
+                                        that were updated in the update_parameters function) and the values of the cost
+                                        function (calculated by the compute_cost function). One value is to be saved
+                                        after each 100 training iterations (e.g. 3000 iterations -> 30 values)..
+        """
+        # initialization
+        parameters = initialize_parameters(layers_dims)
```
Comment (elisim) [inline assignment1/fc_net.py]: we already call this function in __init__
we should think how to organize it

File: assignment1/fc_net.py [hunk 8]
```diff
@@ -99,6 +99,46 @@ def update_parameters(self, parameters, grads, learning_rate):
...
+        costs = []
+
+        for i in range(num_iterations):
+            # choose the batch
+            batch_idx = np.random.choice(np.arange(X.shape[1]), size=batch_size, replace=False)
+            X_batch, Y_batch = X[:, batch_idx], Y[batch_idx]
```
Comment (elisim) [inline assignment1/fc_net.py]: wrong, you can chose the same batch twice. 

it should be something like 

```
def next_batch(X, y, batchSize):
	# loop over our dataset `X` in mini-batches of size `batchSize`
	for i in np.arange(0, X.shape[1], batchSize):
		# yield a tuple of the current batched data and labels
		yield (X[i:i + batchSize], y[i:i + batchSize])
```

File: assignment1/fc_net.py [hunk 9]
```diff
@@ -99,6 +99,46 @@ def update_parameters(self, parameters, grads, learning_rate):
 
         return parameters
 
+    def L_layer_model(self, X, Y, layers_dims, learning_rate, num_iterations, batch_size):
+        """
+
+        :param X: the input data, a numpy array of shape (height*width , number_of_examples)
```
Comment (elisim) [inline assignment1/fc_net.py]: In our implementation we assume the n_examples is the first dimension, no? If yes, we need to reshape

File: assignment1/fc_net.py (hunk 4 above)
Comment (galbraun) [reply assignment1/fc_net.py]: in other places we started with 0

File: assignment1/layers.py (hunk 5 above)
Comment (galbraun) [reply assignment1/layers.py]: look in the first presentation of the course in slide 50

File: assignment1/fc_net.py (hunk 7 above)
Comment (galbraun) [reply assignment1/fc_net.py]: yes i agree, need to be structured better

File: assignment1/fc_net.py (hunk 8 above)
Comment (galbraun) [reply assignment1/fc_net.py]: 👍 will be fixed

File: assignment1/fc_net.py (hunk 9 above)
Comment (galbraun) [reply assignment1/fc_net.py]: we need to be consistent in all of our code you are right
